### DISCUSSION TITLE
How to avoid duplicate components in bundle

### DISCUSSION THREAD

**Discussion by negasora** [bevyengine/bevy#10290] [Q&A] [answered]
Hi! I'm relatively new to bevy and ECS in general - I'm trying to display a player name followed by a health bar, resulting in something looking like `Player: [======]`. But I'm running into a panic because `TextBundle` and `NodeBundle` have duplicate components. How would I go about doing this? I've looked through the unofficial book and can't find anything obvious that would help.

Example Code (using version `0.11.3`):
```rust
use bevy::prelude::*;

#[derive(Bundle)]
struct PlayerHealthBundle
{
    name: TextBundle,
    health_bar: NodeBundle,
}

fn example_startup_system(mut commands: Commands)
{
    commands.spawn(Camera2dBundle::default());

    commands.spawn(PlayerHealthBundle {
        name: TextBundle::from_section(
            "Player: ",
            TextStyle { font_size: 10.0, color: Color::WHITE, ..default() }
        ),
        health_bar: NodeBundle {
            style: Style {
                width: Val::Px(100.0),
                height: Val::Px(20.0),
                ..default()
            },
            background_color: BackgroundColor(Color::GREEN),
            ..default()
        }
    });
}

fn main() {
    App::new()
    .add_plugins(DefaultPlugins)
    .add_systems(Startup, example_startup_system)
    .run();
}
```

**Comment by nicopap** [comment, marked as answer]
In this particular case, **you don't need the `NodeBundle`**, since `TextBundle` already includes all the necessary components.

But for future references, you can't really create a bundle this way, your best bet is to insert the bundles separately, using a chain of `.insert()` as follow:

<details><summary>When you actually need to avoid duplicate components in bundles</summary>

* Issue: https://github.com/bevyengine/bevy/issues/6622

```rust
    commands.spawn(TextBundle::from_section(
        "Player: ",
        TextStyle { font_size: 10.0, color: Color::WHITE, ..default() }
    ))
	.insert(NodeBundle {
        style: Style {
            width: Val::Px(100.0),
            height: Val::Px(20.0),
            ..default()
        },
        background_color: BackgroundColor(Color::GREEN),
        ..default()
    })
```

</details>

**Reply by negasora** [reply]
Sorry, I should've specified: I don't want to draw the health bar as text but as a rectangle that fills/empties as health goes up/down

**Reply by nicopap** [reply]
You need to spawn those as two separate entities. One for the text, and one for the "health bar".

**Reply by negasora** [reply]
Thank you! Spawning as distinct entities with marker components worked great
